File: indexer.py
from reader import Reader
import os, persistence


class Indexer:
    persistence = persistence.JsonDictionaryPersistence()

    _book_names = {}
    _word_indexes = {}

    def _index_one(self, path, index):
        reader = Reader()
        book_name, words = reader.read_book(path)
        self._book_names[index] = book_name

        for word in words:
            if word not in self._word_indexes:
                self._word_indexes[word] = {index}
            else:
                self._word_indexes[word].add(index)

    def index_all(self, directory):
        """
        """
        for i, filename in enumerate(os.listdir(directory)):
            file = os.path.join(directory, filename)
            # checking if it is a file
            if os.path.isfile(file):
                self._index_one(file, i)

    def get_indexes_of_word(self, word):
        return self._word_indexes[word]

    def get_book_name(self, index):
        return self._book_names[index]

    def get_book_names_for_indexes(self, indexes):
        book_names = []
        for index in indexes:
            book_names.append(self._book_names[index])
        return book_names

    def dump(self, book_names_path="book_names.json", word_index_path="word_indexes.json"):
        self.persistence.dump_dict(book_names_path, self._book_names)
        serialized_dict = {key: list(value) for key, value in self._word_indexes.items()}
        self.persistence.dump_dict(word_index_path, serialized_dict)

    def load(self, book_names_path="book_names.json", word_index_path="word_indexes.json"):
        loaded_book_names = self.persistence.load_dict(book_names_path)
        self._book_names = {int(key): value for key, value in loaded_book_names.items()}

        loaded_word_indexes = self.persistence.load_dict(word_index_path)
        self._word_indexes = {key: set(value) for key, value in loaded_word_indexes.items()}

File: java/Indexer/src/main/java/Indexer.java
import datalake.Book;
import datalake.DataLake;
import datalake.Reader;
import impl.DataMart;
import impl.file.FileDataMart;

import java.io.File;
import java.util.List;

public class Indexer {

    private DataMart dataMart;
    private DataLake dataLake;

    public Indexer(DataLake dl) {
        dataMart = new FileDataMart();
        dataLake = dl;
    }

    public void indexOne(String path) {
        Reader reader = new Reader(dataLake);
        Book book = reader.readBook(path);
        int index = book.getIndex();
        String bookName = book.getName();
        List<String> words = book.getWords();
        System.out.println("[INDEXER]: Indexing book: \"" + bookName + "\"");
        for (String word : words) {
            dataMart.addBookIndexToWord(word,index);
        }
    }

    public void indexAll() {
        String directory = dataLake.getDataLakePath();
        System.out.println("[INDEXER]: ------------------ Indexing starting -------------------");
        File[] files = new File(directory).listFiles();
        if (files != null) {
            for (int i = 0; i < files.length; i++) {
                if (files[i].isFile()) {
                    indexOne(files[i].getAbsolutePath());
                }
            }
        }
        System.out.println("[INDEXER]: -------------------- Indexing ended --------------------");
    }

}

